<h1 align="center">DiVRGE<br> <em>Deletions in Viral Genomes Extractor</em> </h1>


<p align="center">
<em>reference in preparation</em><br>
Katherine E. E. Johnson (1), Alexandra A. Mushegian (2), Scott W. Long (3), Randall J. Olsen (3), Paul A. Christensen (3), Matthew Chung (2), James Musser (3), Elodie Ghedin (1,2)*
</p>

1. Center for Genomics and Systems Biology, Department of Biology, New York University, New York, NY, USA.<br>

2. Systems Genomics Section, Laboratory of Parasitic Diseases, National Institute of Allergy and Infectious Disease, National Institutes of Health, Bethesda, MD, USA.<br>

3. Laboratory of Molecular and Translational Human Infectious Diseases Research, Center for Infectious Diseases, Department of Pathology and Genomic Medicine, Houston Methodist Research Institute and Houston Methodist Hospital Houston, Texas, 77030 <br>


<h2 align="center"> Running DiVRGE </h2>

DiVRGE has been tested using alignment outputs generated by BBmap [1] and STAR [2] and will work on any alignment file that contains 'N' to represent a split-read in the CIGAR string [3]. We highly recommend using BBmap as it is an intuitive aligner to work with and is more accurate at aligning split-reads from larger and more complex viral genomes, such as SARS-CoV-2. 

<h2 align="center"> Setup and Requirements </h2> <br>


<p align="center">
After cloning the DiVRGE repo, check that all required programs and packages listed below are installed.<br>
</p>

### Alignment + DiVRGE requirements:
```
trimmomatic/0.36    
bbtools/38.91   
samtools/intel/1.11   
picard/2.23.8   
subread/intel/2.0.1   
python/intel/3.8.6    
```

### DiVRGE Python3 packages (in requirements.txt):
```
Package       Version
------------- ---------
biopython     1.81
joblib        1.2.0
numpy         1.21.5
pysam         0.20.0
scikit-learn  1.2.1
scipy         1.10.0
sklearn       0.0.post1
threadpoolctl 3.1.0
```
### Setting up virtual environment named 'divrge' and installing packages: 
```
>cd $HOME  # select direct. to create environment
>python3 -m venv divrge  # create new python environment
>source divrge/bin/activate  # activate env
>python3 -m pip install -r requirements.txt  # install DiVRGE requirements
```

### Check that necessary Python packages are installed in virtual env. using *pip list --local*:<br> 
```
(divrge) [user ~]$ pip list --local
Package       Version
------------- ---------
biopython     1.81
joblib        1.2.0
numpy         1.21.5
pysam         0.20.0
scikit-learn  1.2.1
scipy         1.10.0
sklearn       0.0.post1
threadpoolctl 3.1.0
```

### Deactivate env when finished:<br>
```
>deactivate 
```

<h2 align="center"> Alignment + SNV + DiVRGE (individual samples) </h2><br>

<h2 align="center"> DiVRGE only (individual samples) </h2><br>

<h2 align="center"> DiVRGE only (grouped samples) </h2><br>



## Contact: 
