<h1 align="center">DiVRGE<br> <em>Deletions in Viral Genomes Extractor</em> </h1>


<p align="center">
<em>reference in preparation</em><br>
Katherine E. E. Johnson (1), Alexandra A. Mushegian (2), Scott W. Long (3), Randall J. Olsen (3), Paul A. Christensen (3), Matthew Chung (2), James Musser (3), Elodie Ghedin (1,2)*
</p>

1. Center for Genomics and Systems Biology, Department of Biology, New York University, New York, NY, USA.<br>

2. Systems Genomics Section, Laboratory of Parasitic Diseases, National Institute of Allergy and Infectious Disease, National Institutes of Health, Bethesda, MD, USA.<br>

3. Laboratory of Molecular and Translational Human Infectious Diseases Research, Center for Infectious Diseases, Department of Pathology and Genomic Medicine, Houston Methodist Research Institute and Houston Methodist Hospital Houston, Texas, 77030 <br>


<h2 align="center"> Running DiVRGE </h2>

DiVRGE has been tested using alignment outputs generated by BBmap [1] and STAR [2] and will work on any alignment file that contains 'N' to represent a split-read in the CIGAR string [3]. We highly recommend using BBmap as it is an intuitive aligner to work with and is more accurate at aligning split-reads from larger and more complex viral genomes, such as SARS-CoV-2. 

<h2 align="center"> Quick Start </h2><br>
1. Clone repo<br>
2. Setup Python Virtual Environment<br>
3. Adjust config file<br>
4. Run snakemake: snakemake -c2 --use-envmodules -s snakefile<br>

<h2 align="center"> Setup and Requirements </h2> <br>


<p align="center">
After cloning the DiVRGE repo, check that all required programs and packages listed below are installed.<br>
</p>

### Alignment + DiVRGE requirements:
```
trimmomatic/0.36    
bbtools/38.91   
samtools/intel/1.11   
picard/2.23.8   
subread/intel/2.0.1   
python/intel/3.8.6    
```

### DiVRGE Python3 packages (in requirements.txt):
```
Package       Version
------------- ---------
biopython     1.81
joblib        1.2.0
numpy         1.21.5
pysam         0.20.0
scikit-learn  1.2.1
scipy         1.10.0
sklearn       0.0.post1
threadpoolctl 3.1.0
```
### Setting up virtual environment named 'divrge' and installing packages: 
```
>cd $HOME  # select direct. to create environment
>python3 -m venv divrge  # create new python environment
>source divrge/bin/activate  # activate env
>python3 -m pip install -r requirements.txt  # install DiVRGE requirements
```

### Check that necessary Python packages are installed in virtual env. using *pip list --local*:<br> 
```
(divrge) [user ~]$ pip list --local
Package       Version
------------- ---------
biopython     1.81
joblib        1.2.0
numpy         1.21.5
pysam         0.20.0
scikit-learn  1.2.1
scipy         1.10.0
sklearn       0.0.post1
threadpoolctl 3.1.0
```

### Deactivate env when finished:<br>
```
>deactivate 
```

<h2 align="center"> Adjusting config and running </h2><br>

### snakefile_main_align (recommended) includes: Align + SNV + DiVRGE (individual samples)<br>

- Download config_align.yaml file
- Adjust using your inputs (see below for details)
- Do not rename
- Works on paired-end seq. data (must adjust for single-end data)

#### config_align.yaml: 
```
---
params:  # general and alignment parameters
  ref: path and ref file used for alignmnet ("/path/to/reference/h9n2/H9N2_HK_1999_CDS.fasta") 
  reads: path to raw fastq data ("/path/to/rawfiles/")
  begin1: characters found before sample id for pair 1 reads (ex: "n01", "000000-ZJFPQ", "HT5MLBGXL_n01_")
  ending1: characters at end of read name, after the sample id for pair 1 reads (ex: "_read1.fq.gz", "fastq.gz", "fq.gz", "n01.fq.gz")
  begin2: characters found before sample id for pair 2 reads (ex: "n02", "000000-ZJFPQ", "HT5MLBGXL_n02_")
  ending2: characters at end of read name, after the sample id for pair 2 reads (ex: "_read2.fq.gz", "fastq.gz", "fq.gz", "n02.fq.gz")
  output: path to directory where files should be saved. MAKE SURE '/' PRESENT AT END. ("/path/to/save/direct/")
  maxindel: the max deletion size expected. BBmap input. (ex: "3000")
  minindel: the min deletion size allowed. Warning: deletion sizes <5nt will decrease accuracy of alignment. BBmap input. (ex: "5")
  adapters: path and filename to adapters used for trimming (ex: "/share/apps/trimmomatic/0.36/adapters/NexteraPE-PE.fa")
  strain: strain, virus, subtype, lineage, etc. for naming purposes (ex: "H1N1","H3N2","SARS2","COVID")
  gigs: the amt. of mem needed for alignment with BBmap. Lg. files will require more memory (ex: "4G", "8G", "16G") 
  gtf_dir: filepath to where gtf file should be saved (ex: "/path/to/reference/h9n2")
  script_dir: directory where 'timo' and 'divrge' scripts are located (ex: "/scratch/kej310/testit/scripts/")
  venv_path: path to python environment with installed packages (see above for more info on installing packages) ("/path/to/divrge/bin/activate")

timo_params:  # timo specific parameters
  segments: list of 'segments' or 'CHROM' in the reference file. Include the brackets. (["H9N2_PB2","H9N2_PB1","H9N2_PA","H9N2_HA","H9N2_NP","H9N2_MP","H9N2_NS"])
  freq_cutoff: min freq of minor SNV required (default: 0.001) (ex: "0.001")
  cov_cutoff: min cov. required for SNV (default: 1) (ex: "1")

divrge_param: # divrge specific parameters
  mismatch: number of mismatches allowed in read alignment (default: "2")
  gap_size: min. size of deletion (default: "5")
  align_length: min. alignment length before and after deletion (default: "28")
  group_size: grouping size (default: "5")
  ncores: number of cores used during grouping (default: "2")
  ninput: number of samples input (default: "1")

mod_params: # modules to load
  subread_mod: "subread/intel/2.0.1"
  samtools_mod: "samtools/intel/1.14"
  python_mod: "python/intel/3.8.6"
  trim_mod: "trimmomatic/0.36"
  trim_jar: "$TRIMMOMATIC_JAR" (path to trimmomatic jar file)
  bbtools: "bbtools/38.91"
  bwa: "bwa/intel/0.7.17"
  picard_mod: "picard/2.23.8"
  picard_jar: "$PICARD_JAR" (path to picard jar file)

```

### snakefile_divrge_only includes: DiVRGE only (individual samples)<br>
### snakefile_divrge_group includes: DiVRGE (grouped samples)<br>


## Contact: 
